Document napari plugin (#393)
neuroinformatics-unit/movement: docs/source/user_guide/napari_plugin.md

(target-napari-plugin)=
# The napari plugin

To enable the interactive visualisation of motion tracks in

> `movement`, we have created an experimental plugin for [napari](napari:).

Author: It's a bit more complicated. In napari-lingo, plugins are extensions of the core napari functionality, and they may (or may not) come with widgets. For example `napari-video` doesn't have a widget, it's purely a so-called "reader" plugin.

I think I will go with sth like this:

```md
You can visualise `movement` motion tracks in [napari](napari:) using our plugin.
Plugins can extend `napari`'s core functionality by adding
features like file format support, custom data processing, and widgets
that can be docked as panels in the `napari` window.

Our `movement` plugin for `napari` currently allows you to visualise 2D
[poses datasets](target-poses-and-bboxes-dataset)
as points overlaid on video frames.
```

> `movement`, we have created an experimental plugin for [napari](napari:).

Author: Actually, I decided to put it more succintly (it felt weird to go on long tangents about what's a plugin and what's a widget).

```md
You can visualise `movement` motion tracks in [napari](napari:) using our
plugin, which extends `napari`'s core functionality and provides a widget
that can be docked in the `napari` window. Currently, you can use it to
visualise 2D [poses datasets](target-poses-and-bboxes-dataset)
as points overlaid on video frames.
```

Currently, the plugin supports loading 2D
[poses datasets](target-poses-and-bboxes-dataset), and visualising them
as points overlaid on video frames.

:::{warning}
This plugin is still in early stages of development and offers
limited functionality. We are working on ironing out the kinks and
gradually adding more features. [Get in touch](target-get-in-touch)
if you find any bugs or have suggestions for improvements!
:::

## Installation

The `napari` plugin is shipped with the `movement` package starting from

> version `0.1.0`. If you install `movement` via `conda`, `napari` is already
> included as a dependency. If you use the `pip` installer, make sure to

Reviewer: ```suggestion
version `0.1.0`. If you install `movement` via `conda`, `napari` will already
be available. If you use the `pip` installer, make sure to
```

For the users that are less familiar with the dev terms?

install `movement` with the `[napari]` extra:

```sh
pip install movement[napari]
```

> ## Launch the plugin

Author: I will go with "Launch napari"

> ## Launch the plugin

Author: I went with "Launch napari"

Type the following command in your terminal:

```sh
napari -w movement
```

This will open the `napari` window with the `movement` plugin docked on the

> right-hand side.

Author: good point, will do.

> 

Reviewer: Maybe just before this section we can introduce the concept of layer and the layer list. Something like:
```suggestion
In napari, data is typically loaded into layers, which can be reordered and toggled for visibility in the layers list panel. For example, keypoint data can be added as a 'points' layer, while images or videos can be loaded as 'image' layers. Below, we’ll explain how to do this.
```

Reviewer: It may be important to say the order affects how the data is presented (e.g., if the video is the top layer it covers the keypoints). 

We may also want to mention one can load multiple videos and pose data files, and toggle their visibility. Maybe people use that?

Author: I added the text you suggested, which is already very helpful. I also cross-referenced the word "layers" to napari's [layers at a glance](https://napari.org/0.5.3/guides/layers.html) guide.

## Load a background layer

> Next, you need a background for you visualisation. You can either

Author: will do.

[load the video](target-load-video) corresponding to the poses dataset,
or a [single image](target-load-frame), e.g., a still frame
derived from that video. In the following sections, we will show you how to
do both and discuss some limitations.

(target-load-video)=
### Load a whole video

> To load a video, drag and drop the video file onto the `napari` window.

Author: Great idea! I kept wishing drag-and-dropping would work while I was testing the plugin myself. Is saves so much time!

I opened issue #422 to track this.

You will see a pop-up dialog asking you to select the reader.
Choose the `video` reader—corresponding to the
[`napari-video`](https://github.com/janclemenslab/napari-video)
plugin—and click `OK`.

`napari-video` will load the video as a single `napari`
[image layer](napari:howtos/layers/image.html), with a slider
at the bottom that you can use to navigate through frames.
You may also use the left and right arrow keys to navigate
frame-by-frame.

Clicking on the play button will start the video playback at a default

> rate of 10 frames per second. You can adjust that by right-clicking on the

Author: Actually, I think there is an (experimental) way to do this: https://napari.org/dev/gallery/viewer_fps_label.html
Good idea, I didn't know this was possible.

> rate of 10 frames per second. You can adjust that by right-clicking on the

Author: I will give it a shot.

> rate of 10 frames per second. You can adjust that by right-clicking on the

Author: There seems to be an experimental way of doing this in napari: https://napari.org/dev/gallery/viewer_fps_label.html

> rate of 10 frames per second. You can adjust that by right-clicking on the

Author: This turned out to be a bit of a red herring 😢 
The example I linked above tries to measure the "actual" fps in real time, which fluctuates wildly and usually hovers around the selected playback fps value. This is more confusing than helpful. Plus it relies on accessing a forbidden private attribute and throws all sorts of warnings.

I next tried to display the "playback fps setting" instead of the "actual" measured fps, but I spend lots of time trying to intercept the right napari events to trigger the update of that display. I will table that for now in the interest of not  going down a massive rabbithole.

> rate of 10 frames per second. You can adjust that by right-clicking on the

Reviewer: yes sounds good, sorry to send you down there with all them rabbits 🐇 ! 

I will open an issue so we have somewhere to go to if we pursue this further some day. I will paste these thoughts if that is alright?

> rate of 10 frames per second. You can adjust that by right-clicking on the

Author: Of course, thank you!

> rate of 10 frames per second. You can adjust that by right-clicking on the

Reviewer: done --> #423

play button or by opening the `napari > Preferences` menu and changing
the `Playback frames per second` setting.

:::{admonition} Video playback limitations
:class: warning

> - During playback you cannot jump at an arbitrary frame by clicking on the

Author: If nothing bad happened you are lucky. I tried it on different machines and on many of them it "stutters", "freezes" or there is just a big delay to switch to the new frame. Here we don't care much about retrieving the correct frames, it's more that it leads to a bad user experience.

Maybe I reward it to "Refrain from jumping" instead of "you cannot jump"?

> - During playback you cannot jump at an arbitrary frame by clicking on the

Reviewer: Or maybe just as you explained it: if you dont stop it before jumping to a new frame, the playback may freeze or stutter.

> - During playback you cannot jump at an arbitrary frame by clicking on the

Author: How about this?

"The video playback may freeze or stutter if you click on the slider to jump to a specific frame. We recommended pausing the playback before such jumps."

  using a [single image](target-load-frame) as a background.
:::

(target-load-frame)=
### Load an image

This usually means using a still frame extracted from the video, but in theory
you could use any image that's in the same coordinate system as the
tracking data. For example, you could use a schematic diagram of the arena,
as long as it has the same width and height as the video and is
properly aligned with the tracking data.

::: {dropdown} Extracting a still frame from a video
:color: info
:icon: info

You can extract a still frame using your video player/editor of
choice. We find the command line tool [`ffmpeg`](https://www.ffmpeg.org/)
very useful for this task.

To extract the first frame of a video:

```sh
ffmpeg -i video.mp4 -frames:v 1 first-frame.png
```

To extract a frame at a specific time stamp (e.g. at 2 seconds):

```sh
ffmpeg -i video.mp4 -ss 00:00:02 -frames:v 1 frame-2sec.png
```
:::

To load any image into `napari`, simply drag and drop the image file into
the napari window. Alternatively, you can use the `File > Open File(s)` menu
option and select the file from the file dialog.
In any case, the image will be loaded as a static

> [image layer](napari:howtos/layers/image.html).

Author: Yeah I will try to reword this to avoid confusion. But I want to keep the references to the image layer docs, because users can easily refer to them to learn the full features of napari and its layers.

There are many cool things one can do in napari but I didn't want to basically replicate their docs, that's why I tried to sprinkle in references to the most relevant places in their docs.

## Load the poses dataset

Now you are ready to load some pose tracks over your chosen background layer.

> The `movement` plugin on the right-hand side of the window should contain
> an expanded `Load poses` menu. This contains a `source software` dropdown,
> an `fps`  (frames per second) input field, and a `Browse` button to select
> a file containing predicted poses.
> You may also directly paste a `file path` into the homonymous field.

Reviewer: We may want to express this as a series of steps, rather than a description of the UI.

Something like:
```suggestion
On the right-hand side of the you should see
an expanded `Load poses` menu. To load some pose data in napari:
1- Select the `source software` from the dropdown menu
2- Set the `fps`  (frames per second) of the video the pose data refers to. Note this will only affect the units of the time variable shown when hovering over a keypoint. To show the units in frames, or if the fps is not known, it can be set to 1.
3 - Select the file containing the predicted poses. The path can be directly pasted or you can use the file browser button.
4 - Click Load

The data should be loaded into the viewer as a [points layer](napari:howtos/layers/points.html). By default, it is added at the top of the layer list.
```

Reviewer: Spelling it out could minimise errors due to file and format mismatches.

Author: This is much much better! Thanks ❤️

::: {note}
See [supported formats](target-supported-formats) for more information on
the expected software and file formats.
:::

> Once the file path is set, click the `Load` button to load the poses dataset
> into `napari`. The plugin will create a new
> [points layer](napari:howtos/layers/points.html) on top of the existing
> image layer.

Reviewer: ```suggestion
```
If you take the suggestion above, this could be removed.

You will see a view similar to the one below:

![napari plugin with poses dataset loaded](../_static/napari_plugin_with_poses_as_points.png)

The predicted keypoints are represented as points, colour-coded by
keypoint ID for single-individual datasets, or by individual ID for

> multi-individual datasets. Hovering with your mouse over a point will

Reviewer: ```suggestion
multi-individual datasets. Hovering with your mouse over a point with the points layer selected will
```
Otherwise the tooltip doesn't work.

bring up a tooltip containing the names of the individual and keypoint,
the point-wise confidence score (as predicted by the source software),
and the time in seconds (this is calculated from the frame number and
the `fps` value you provided).

> Using the slider at the bottom of the window, you can navigate through
> the frames of the poses dataset. The points will update accordingly, allowing
> you to inspect the predicted keypoints at different time points.
> If you loaded a video as a background layer, the video will also update
> in sync with the poses dataset.

Reviewer: ```suggestion
Using the slider at the bottom of the window, you can move through
the frames of the dataset, and the points and video will update
in sync.
```
Just a bit shorter.

::: {admonition} Stay tuned

> Though the display style of the points layer is currently fixed, we are
> working on adding more customisation options in future releases, such as
> enabling you to change the point size, colour, or shape.

Reviewer: ```suggestion
We are working on adding more customisation options to the display style of the keypoints, such as
changing the point size, colour or shape across all frames.
```
You can change it per frame if you select them all - not too bad! And the opacity does apply across all frames.

I think it is worth mentioning the checkboxes to "display text" and "out of slice" (this one a bit less useful, I think shows a few frames before and after the current one?).

Author: I personally think "out of slice" is confusing and not particularly useful in our use-case, so I don't want to advertise it. "Display text" is a bit more useful, so I will mention it.

```sh
pip install movement[napari]
```

## Launch napari

Type the following command in your terminal:

```sh
napari -w movement
```

This will open the `napari` window with the `movement` widget docked on the
right-hand side, as in the the [screenshot](target-widget-screenshot) below.

In `napari`, data is typically loaded into [layers](napari:guides/layers.html),
which can be reordered and toggled for visibility in the layers list panel.
For example, keypoint data can be added as a
[points layer](napari:howtos/layers/points.html),
while image stacks (including videos) can be added as
[image layers](napari:howtos/layers/image.html).
Below, we'll explain how to do this.

## Load a background layer

Though this is not strictly necessary, it is usually informative to
view the keypoints overlaid on a background that provides
some spatial context. You can either [load the video](target-load-video)
corresponding to the poses dataset, or a [single image](target-load-frame),
e.g., a still frame derived from that video. In the following sections,
we will show you how to do both and discuss some limitations.

(target-load-video)=
### Load a video

> To load a video, drag and drop the video file onto the `napari` window.

Author: Good point, I mention this now in the common section, as you suggested.

```md
## Load a background layer

Though this is not strictly necessary, it is usually informative to
view the keypoints overlaid on a background that provides
some spatial context. You can either [load the video](target-load-video)
corresponding to the poses dataset, or a [single image](target-load-frame),
e.g., a still frame derived from that video. 
You can do this by dragging and dropping the corresponding file onto the
`napari` window or by using the `File > Open File(s)` menu option.
Please read the following sections for detailed information
and some important considerations.
```

  using a [single image](target-load-frame) as a background.
:::

(target-load-frame)=
### Load an image

This usually means using a still frame extracted from the video, but in theory
you could use any image that's in the same coordinate system as the
tracking data. For example, you could use a schematic diagram of the arena,
as long as it has the same width and height as the video and is
properly aligned with the tracking data.

::: {dropdown} Extracting a still frame from a video
:color: info
:icon: info

You can use the command line tool [`ffmpeg`](https://www.ffmpeg.org/)
to extract a still frame from a video.

To extract the first frame of a video:

```sh
ffmpeg -i video.mp4 -frames:v 1 first-frame.png
```

To extract a frame at a specific time stamp (e.g. at 2 seconds):

```sh
ffmpeg -i video.mp4 -ss 00:00:02 -frames:v 1 frame-2sec.png
```
:::

To load any image into `napari`, simply drag and drop the image file into
the napari window. Alternatively, you can use the `File > Open File(s)` menu
option and select the file from the file dialog.
In any case, the image will be loaded as a single 2D frame without a slider.

> ## Load the poses dataset

Author: I agree that it makes some sense to do it this way, given that only the poses are strictly required. That said, viewing the keypoints without any background is quite uninformative (I haven't found any use for it), so in 90-99% of the cases I expect people to use a background layer. That's why I think it's better to stick to this order, because it's the usual order of operations.

Now you are ready to load some pose tracks over your chosen background layer.

On the right-hand side of the window you should see
an expanded `Load poses` menu. To load some pose data in napari:
1. Select the `source software` from the dropdown menu.
2. Set the `fps`  (frames per second) of the video the pose data refers to. Note this will only affect the units of the time variable shown when hovering over a keypoint. If the fps is not known, you can set it to 1, which will effectively make the time variable equal to the frame number.
3. Select the file containing the predicted poses. The path can be directly pasted or you can use the file browser button.
4. Click `Load`.

The data should be loaded into the viewer as a
[points layer](napari:howtos/layers/points.html).
By default, it is added at the top of the layer list.

::: {note}
See [supported formats](target-supported-formats) for more information on
the expected software and file formats.
:::

You will see a view similar to the one below:

(target-widget-screenshot)=

![napari widget with poses dataset loaded](../_static/napari_plugin_with_poses_as_points.png)

> The predicted keypoints are represented as points, colour-coded by

Reviewer 2: ```suggestion
The keypoints are represented as points, colour-coded by
```

keypoint ID for single-individual datasets, or by individual ID for
multi-individual datasets. These IDs can be also displayed as text
next to the points by enabling the `display text` option from the
layer controls panel.

Hovering with your mouse over a point
(with the points layer selected) will
bring up a tooltip containing the names of the individual and keypoint,
the point-wise confidence score (as predicted by the source software),

> and the time in seconds (this is calculated from the frame number and

Reviewer 2: ```suggestion
and the time in seconds (calculated based on the frame number and
```
